src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs

```diff
@@ -2676,5 +2676,10 @@ private void Remove(int startIndex, int count, out StringBuilder chunk, out int
             Debug.Assert(chunk != null, "We fell off the beginning of the string!");
             AssertInvariants();
         }
+
+        private static char[] AllocateCharArray(int size)
+            => size > 85_000 / sizeof(char)
+                ? GC.AllocateUninitializedArray<char>(size) // lacks of zeroing gives a nice perf boost for LOH-allocated arrays
+                : new char[size]; // small arrays very often come from a pre-zeored thread allocation context, calling AllocateUninitializedArray would decrease the perf
```

Comment:
Have you observed that it is actually decreasing the perf?

`AllocateUninitializedArray` has a fast path for small arrays very much like this, just the threshold is lower. If the implementation of `AllocateUninitializedArray` is not tuned correctly, we should fix that inside `AllocateUninitializedArray`. It does not make sense everybody wrapping `AllocateUninitializedArray` with a boiler plate code like this.

cc @VSadov

Reply:
I have opened https://github.com/dotnet/coreclr/issues/27203 so that this is not forgotten. I agree that the fast path threshold in AllocateUninitializedArray always felt way too low.

Reply:
Yes. The threshold is most certainly too low for StringBuilder. 1 - 2 K may work better.

The issue with UninitializedArrays is that performance depends on the call pattern.
There is more benefit if you call this repeatedly. If you make a normal allocation than the benefit is diminished. 

A normal allocation of size X will leave from 0 to (4K - X)  pre-cleared budget. The next allocation within this budget will not see benefits of Uninitialized it fits.

If you mostly do normal allocations and they are relatively small, and then occasionally do Uninitialized, then threshold should be in the order of 2K. Since that is the expected average budget.
On the other hand - If you call the API repeatedly in a loop, then you want smaller threshold since yor cost is dominated by memclearing. 

So far the only use of the API is when ArrayPool rent is called repeatedly (without returning). The threshold is  tuned for the only user. 

I think,  the pattern used by ArrayPool oversubscription benchmark is not common. **I feel for users like StringBuilder the threshold should be about 1 - 2K .**
Also one number would not fit all. What will be good for StringBuilder may cause regression for ArrayPool. 
And I think that is OK, since the ArrayPool oversubscription, while possible, is basically an anomaly. We should handle it reasonably well, but should not optimize for it at the cost to other users.

Reply:
I'll add this to the bug. It may be a good time to add more users to the Uninitialized API and tune the threshold.